Universal Assisted Generation: Assisted generation with any assistant model (by Intel Labs)

danielkorat opened the pull request:
# What does this PR do?
Co-Authors: @mosheber, @jmamou, @orenpereg

Assisted Generation (AG; i.e. Speculative Decoding) can accelerate inference speed by using a small assistant model in tandem with the original (target) model. However, it only supports assistant models that use the same tokenizer as the target model. This significantly limits the usability of AG, as only few models have such corresponding assistant models.
This PR alleviates that requirement, allowing AG execution with any assistant model regardless of its tokenizer.
Thus, AG can support any target model (!), allowing acceleration of a wide range of models that was not possible until now.

For example, our initial experiments show the following speedups:

| target | assistant | dataset | task | speedup |
|----------------------|---------------------|---------------------------|---------------------------|---------------------------|
| `codellama/CodeLlama-13b-Instruct-hf` | `bigcode/tiny_starcoder_py` | `openai/humaneval` | code generation | **2.06x** |
| `microsoft/Phi-3-medium-128k-instruct` | `Qwen/Qwen2-0.5B-Instruct`  | `tau/scrolls`   | long-context summarization | **1.63x** |
| `google/gemma-2-9b` | `double7/vicuna-68m`  | `cnn_dailymail`   | summarization | **1.90x** |

Our method works by decoding the speculated tokens and re-encoding them using the target tokenizer, and then performing an alignment procedure, to account for inter-tokenizer discrepancies.

Usage example:
```python
out = model.generate(input_ids, do_sample=False, assistant_model=assistant_model, 
                                  assistant_tokenizer=assistant_tokenizer, tokenizer=target_tokenizer
)
```

where `target_tokenizer` and `assistant_tokenizer` are the respective tokenizers; `target_lookbehind` and `assistant_lookbehind` are optional parameters which define the window size to consider for solving inter-tokenizer discrepancies after every generation step of the assistant model. 

Experimental setup:
1 x A6000 GPU
100 examples per dataset

@gante

CC @lewtun

## Before submitting
- [x] <s>This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).</s>
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] <s>Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.</s>
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

@gante (all others)
CC @lewtun 

 -->

danielkorat committed:
Update candidate_generator.py

danielkorat committed:
Update utils.py

danielkorat committed:
add lookbehind params to _get_candidate_generator

danielkorat committed:
make fixup

danielkorat committed:
add unit tests

danielkorat committed:
fix failing tests

danielkorat committed:
add docstrings

danielkorat committed:
fix docstrings; remove non-optimized AnyTokenizer

mosheber committed:
added any tokenizer generation correctness test

danielkorat commented:
Note:
The usage can be streamlined by creating a new config class e.g. `SpeculativeConfig`, which initializes the tokenizers and has good default values for the lookbehinds:

```python
sd_config = SpeculativeConfig(model_name_or_path=model_name_or_path, assistant_name_or_path=assistant_name_or_path)

out = model.generate(input_ids, do_sample=False, max_new_tokens=max_new_tokens, sd_config=sd_config)
```

mosheber committed:
Merge pull request #2 from danielkorat/sd-any-tokenizer-generate-test-basic-1

added any tokenizer generation correctness test

jmamou commented:
> Quick question before I dive into the code: I see @jmamou as a co-author, is this PR related to DISCO (and, therefore, to #33258 ?)

no ... it is orthogonal

danielkorat committed:
Merge branch 'huggingface:main' into sd-any-tokenizer

danielkorat committed:
make fixup

danielkorat committed:
fix assertion syntax

danielkorat committed:
Merge branch 'huggingface:main' into sd-any-tokenizer

danielkorat commented:
Hi @gante 
Our team greatly values your SD and AG implementations in HuggingFace! 😃 
We developed this feature because we wanted to solve one of the main pain points of Speculative Decoding, which does not support model pairs from different families.
Our feature is highly desirable by the community, since it supports acceleration of models which were not supported until now. As an example, our feature is requested in vLLM: [link](https://github.com/vllm-project/vllm/issues/7252).

danielkorat commented:
@gante 

We even got a significant speedup using a 9 billion target model (`google/gemma-2-9b`):
| target | assistant | dataset | task | speedup |
|----------------------|---------------------|---------------------------|---------------------------|---------------------------|
| `codellama/CodeLlama-13b-Instruct-hf` | `bigcode/tiny_starcoder_py` | `openai/humaneval` | code generation | **2.01x** |
| `microsoft/Phi-3-medium-128k-instruct` | `Qwen/Qwen2-0.5B-Instruct`  | `tau/scrolls`   | long-context summarization | **1.65x** |
| `google/gemma-2-9b` | `double7/vicuna-68m`  | `cnn_dailymail`   | summarization | **1.72x** |

gante commented:
@danielkorat @jmamou apologies for the delayed review, bugfixes in the pipeline 👼 I had a quick look at the PR and it looks really really cool, thank you for proposing this technique and opening the PR 💛 

My immediate comment goes in the same direction as a comment added above, about having a ~`SpeculativeConfig`~ `AssistantConfig` to parameterize the added technique, living inside `GenerationConfig`. I am suggesting naming it after assisted generation simply because of consistency throughout the library. Let's please add it, and avoid adding new arguments to `generate` 🤗 This has a few immediate advantages:
1. simpler `generate` signature (advanced users will look into `GenerationConfig`, basic users will appreciate not having to read more flags)
2. we gain the ability to serialize a `GenerationConfig` with a `AssistantConfig` to automatically enable your technique
3. In a future PR, I am thinking of moving all flags related to assisted generation there, to add some organization to the large number of flags :)

gante commented:
I think I will have to work next on `torch.compile` + assisted generation, to squeeze the most of these speedups 😈

gante commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+            new_cur_len = assistant_input_ids.shape[-1]
+        else:
+            # input_ids contains all target prompt input ids and some new target input ids
+            if self.prev_target_ids.shape[1] > self.target_lookbehind:
```
Sanity check: you have confirmed that the runtime with this logic is faster than simply re-encoding the whole `input_ids`?

If it is just a tiny bit faster, like 5%, then I'd prefer having the simpler reencoding code path exclusively -- it's much simpler to maintain and document

(I'll likely have a few comments to make this section more beginner-friendly, but I'm holding them until we confirm this part :D)

danielkorat replied:
Yes, we have tested both options and this one is significantly faster.
Natuarlly, as the number of new tokens increases it becomes even more significant.

danielkorat replied:
In one test we did, it was 15% faster. And that test used `cnn_dailymail` which does not even have long prompts.

Re-encoding (decode+encode) the whole `input_ids` (which increase in size at at each step) at every assistant generation step incurs a significant overhead.

gante replied:
awesome, thank you for confirming

(going to leave this comment unresolved for future reference)

gante commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+
+            new_cur_len = assistant_input_ids.shape[-1]
+
+        max_new_tokens = int(self.num_assistant_tokens)
+
+        min_new_tokens = max(min(max_new_tokens, self.main_model_min_length - new_cur_len), 0)
+        if max_new_tokens == 0:
+            return input_ids, None
```
We can move this block (except `min_new_tokens = ...`) to the top of `get_candidates`, so we can hit this return more quickly 😎

gante commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        assistant_output = self.assistant_model.generate(**assistant_generation_kwargs, **self.assistant_kwargs)
+
+        num_prev_assistant = self.prev_assistant_ids.shape[1]
+        start_assistant_look_index = num_prev_assistant - self.assistant_lookbehind
```
Same question as above: is this code block faster than a naive re-encoding?

gante commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        self.prev_tokens = assistant_output.sequences
+
+        # 4. Prepare variables for output
+        if input_ids.shape[1] >= new_target_ids.shape[1]:
```
In this (corner) case, the assistant model is about to return fewer tokens than it received as input. In the main generation loop, we simply want the main model to ignore the candidate sequences and generate 1 new token.

We can achieve this if we do `return input_ids, None`, as opposed to appending dummy tokens for the main model to reject 🤗 This would be the same as the return above

danielkorat replied:
good catch, thanks !

gante commented on src/transformers/generation/utils.py:
```diff
@@ -3975,8 +4007,13 @@ def _assisted_decoding(
             #  1. Fetch candidate sequences from a `CandidateGenerator`
-            candidate_input_ids, candidate_logits = candidate_generator.get_candidates(input_ids)
-            candidate_input_ids = candidate_input_ids.to(self.device)
+            if isinstance(candidate_generator, AssistedCandidateGeneratorDifferentTokenizers):
+                candidate_input_ids, candidate_logits = candidate_generator.get_candidates(
+                    input_ids, stopping_criteria
+                )
+            else:
+                candidate_input_ids, candidate_logits = candidate_generator.get_candidates(input_ids)
+
```
I think we can avoid these changes: you are passing `stopping_criteria` to reach its `max_length` attribute.

If I'm not mistaken, the assistant model's `generation_config` should the same value for `max_length`

danielkorat replied:
good catch

gante reviewed (commented):
A few comments and suggestions 🤗 

There is a question I want you folks to check first, before all others. I've tagged it with `❗ ❗`

danielkorat commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -28,6 +30,90 @@
     from .configuration_utils import GenerationConfig
 
 
+def _get_longest_diag_dict(some, some_nonzero):
```
sure

danielkorat committed:
PR review fixes

danielkorat committed:
address additional PR comments

danielkorat committed:
fix tests

danielkorat committed:
remove stropping criteria arg

danielkorat committed:
make fixup

danielkorat committed:
add AssistantConfig

danielkorat commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        target_tokenizer (`AutoTokenizer`, *optional*):
+            The tokenizer used for the target model.
+        target_lookbehind (int, *optional*):
+            The number of tokens to look behind in the target model.
```
I will add:
"Since re-encoding the tokens may result in token discrepancies, we need to define a look behind which will mark the start index for finding the matching overlap between the 2 tokenizers, so we can ensure the correct suffix is attached to the `input_ids`."

danielkorat commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+                    assistant_input_ids = torch.cat([assistant_input_ids, new_assistant_ids], dim=-1)
+
+                self.prev_assistant_ids = assistant_input_ids
+            else:
+                assistant_input_ids = self.convert_token_ids(input_ids, **convert_kwargs)
+                assistant_input_ids = torch.cat(
+                    [self.prev_assistant_ids, assistant_input_ids[:, self.prev_assistant_ids.shape[1] :]], dim=1
+                )
+
+            new_cur_len = assistant_input_ids.shape[-1]
```
agreed

danielkorat replied:
agreed

danielkorat committed:
fix prev_tokens branching

danielkorat commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        else:
+            # input_ids contains all target prompt input ids and some new target input ids
+            if self.prev_target_ids.shape[1] > self.target_lookbehind:
+                start_index_in_target_window = self.prev_target_ids.shape[1] - self.target_lookbehind
```
Trying to fix it to 20 and testing it

danielkorat replied:
Currently fixed to 10 and had no issues when running the experiments in the [PR description](https://github.com/huggingface/transformers/pull/33383#issue-2513354628) (50 examples each).

danielkorat committed:
pass tokenizers through `generate()`kwargs

danielkorat committed:
fix lookbehind values; tokenizer params WIP

danielkorat committed:
fixup

danielkorat commented:
@gante I have addressed your comments, feel free to review. 😃
Regarding `AssistantConfig`: I implemented it as you suggested, but it causes a huge slowdown in performance.
I suspect that since it holds two tokenizer objects, and lives inside `Generation Config`, this causes some overhead. I'm still trying to mitigate this. I'm open to any suggestion / solution

danielkorat committed:
AssistantConfig

danielkorat commented:
@gante 
Update:
I managed to find a workaround for the slowdown bug. In `AssistedCandidateGeneratorDifferentTokenizers.get_candidates()`, I remove `assistant_config` from `assistant_generation_kwargs['generation_config']`.

```python
        # 2. Forecast next N tokens using the assistant model.
        assistant_generation_kwargs = {
            self.input_ids_key: assistant_input_ids,
            "min_new_tokens": min_new_tokens,
            "max_new_tokens": max_new_tokens,
            "generation_config": self.generation_config,
            "logits_processor": self.logits_processor,
        }

        self.assistant_kwargs.pop("attention_mask", None)

        assistant_config = self.generation_config.__dict__.pop("assistant_config", None)
        assistant_output = self.assistant_model.generate(**assistant_generation_kwargs, **self.assistant_kwargs)
        self.generation_config.assistant_config = assistant_config
```

 This mitigates the issue. Please advise if you think there's a better solution. I have not pushed this code yer.

danielkorat commented:
@gante 
Another workaround is to pass only `tokenizer` and `assistant_tokenizer` directly to `generate()` using `kwargs`.
Passing `tokenizer` to `generate` is already implemented in `generate`:

https://github.com/huggingface/transformers/blob/2e24ee4dfa39cc0bc264b89edbccc373c8337086/src/transformers/generation/utils.py#L1803

So really this only means adding the option of passing a single additional parameter through `kwargs` to `generate()`, which is `assistant_tokenizer`.

WDYT?

danielkorat commented:
@gante 
After running some tests, this line is causing the slowdown mentioned above:

https://github.com/huggingface/transformers/blob/f5247aca0151de9b53096772d3ba0516ee1b896f/src/transformers/generation/candidate_generator.py#L158

The line above is super slow since it copies two Tokenizer objects living inside `generation_config.assistant_config`.

A proper solution could be implementing something like this for `GenerationConfig`:

```python
from copy import deepcopy

class Foo:
    def __init__(self, content, linked_to):
        self.content = content
        self.linked_to = linked_to

    def __deepcopy__(self, memo):
        # create a copy with self.linked_to *not copied*, just referenced.
        return Foo(deepcopy(self.content, memo), self.linked_to)
```

gante commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -456,6 +456,9 @@ def __init__(self, **kwargs):
         self.num_assistant_tokens_schedule = kwargs.pop("num_assistant_tokens_schedule", "heuristic")
         self.assistant_confidence_threshold = kwargs.pop("assistant_confidence_threshold", None)
 
+        self.target_tokenizer = kwargs.pop("target_tokenizer", None)
+        self.assistant_tokenizer = kwargs.pop("assistant_tokenizer", None)
```
Let's try to find a solution where the tokenizers are not inside the generation config. Even if it means extra kwargs in `generate` (we should avoid complex objects in the config instances)

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1154,7 +1175,7 @@ def _validate_assistant(self, assistant_model):
 
-        if not self.config.vocab_size == assistant_model.config.vocab_size:
+        if assistant_tokenizer is None and self.config.vocab_size != assistant_model.config.vocab_size:
             raise ValueError("Make sure the main and assistant model use the same tokenizer")
```
Can we modify this message to suggest using your technique? (i.e. to pass the assistant tokenizer as well 🤗 )

gante commented on tests/generation/test_utils.py:
```diff
@@ -3493,3 +3494,77 @@ def test_generate_from_inputs_embeds_with_bos_token_id_is_none(self):
             model.generate(max_length=20, bos_token_id=None)
+
+
+class TestGetTokensDiag(unittest.TestCase):
```
Nit: can we rename it to something like `TestAssistedCandidateGeneratorDifferentTokenizers` (so we know all tests here are related to `AssistedCandidateGeneratorDifferentTokenizers`)?

gante commented on tests/generation/test_utils.py:
```diff
@@ -3493,3 +3494,77 @@ def test_generate_from_inputs_embeds_with_bos_token_id_is_none(self):
+
+
+class TestAnyTokenizerGenerationCorrectness(unittest.TestCase):
+    def test_speculative_decoding_equals_regular_decoding(self):
```
Nit: let's move this to the integration tests above (`GenerationIntegrationTests`). We place all `generate` integration tests there 🤗

gante reviewed (commented):
Aside from the tokenizer placement in the generation config, LGTM 🤗

danielkorat committed:
remove AssistantConfig; apply PR suggestions

danielkorat committed:
restructure tests

danielkorat committed:
fixup

danielkorat committed:
fix assistant_tokenizer arg validation

danielkorat committed:
Merge branch 'main' into sd-any-tokenizer

danielkorat committed:
fixup

danielkorat committed:
fix tests in TestAssistedCandidateGeneratorDifferentTokenizers

danielkorat commented:
Thanks @gante, applied your suggestions 😃 

Ready to merge (failing tests are unrelated to this PR)

gante reviewed (approved):
Perfect! Thank you for iterating 🤗

danielkorat committed:
Merge branch 'huggingface:main' into sd-any-tokenizer

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33383). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

danielkorat committed:
fix class docstring

gante commented:
(the broken CI should be fixed by #33950 , let's wait for it to be merged to then rebase :) )

ArthurZucker reviewed (commented):
Very nice addition! Needs a tad bit more documentation!

danielkorat commented on tests/generation/test_utils.py:
```diff
@@ -3507,6 +3508,34 @@ def test_special_tokens_fall_back_to_model_default(self):
         self.assertTrue(generation_config.bos_token_id is None)
 
+    def test_speculative_decoding_equals_regular_decoding(self):
+        draft_name = "double7/vicuna-68m"
```
@ArthurZucker which doc do you mean here?

ArthurZucker replied:
I mean the doc that is missing from this PR 😉

danielkorat commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -258,6 +261,280 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        return dest_ids.to(input_ids.device)
+
+    def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor, Optional[torch.FloatTensor]]:
+        max_new_tokens = int(self.num_assistant_tokens)
```
Great idea, will add an explanation and code example under: 
https://github.com/huggingface/transformers/blob/330bb6b3847ff91eacfe00842545a7c9c9eb51a7/docs/source/en/generation_strategies.md#L408
After the blog is released, will also link to the blog here and in `generation_strategies.md`

ArthurZucker replied:
Sounds good

danielkorat committed:
PR suggestions

danielkorat committed:
doc

danielkorat committed:
doc update and improvements to `_validate_assistant()`

danielkorat commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -258,6 +261,280 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        Returns:
+            The converted token IDs.
+        """
+        text = source_tokenizer.batch_decode(input_ids, skip_special_tokens=True, clean_up_tokenization_spaces=True)
```
Thanks @ArthurZucker 
This line ensures more consistent re-encoding.
I just tested with `lean_up_tokenization_spaces=False` and the speedup was indeed a bit lower.

ArthurZucker replied:
alright thanks!

danielkorat commented:
> (the broken CI should be fixed by #33950 , let's wait for it to be merged to then rebase :) )

@ArthurZucker Since that PR is taking time, any chance we can merge this PR without having to wait?

danielkorat committed:
Merge branch 'huggingface:main' into sd-any-tokenizer
